src/Futhark/Pass/OptimizeArrayLayout/Layout.hs

```diff
@@ -0,0 +1,242 @@
+      | UnOpExp (Complement _) _ <- op = Nothing
+      | UnOpExp (Abs _) _ <- op = Nothing
+      | UnOpExp _ sub_op <- op = reduceStrideAndOffset sub_op
+      | ConvOpExp _ sub_op <- op = reduceStrideAndOffset sub_op
+      | otherwise = Nothing
+reduceStrideAndOffset _ = Nothing
+
+multicorePermutation :: PrimExpTable -> SegOpName -> ArrayName -> IndexExprName -> [DimAccess rep] -> Maybe Permutation
+multicorePermutation primExpTable _segOpName (_arr_name, nest, arr_layout) _idx_name dimAccesses = do
+  -- Create a candidate permutation
+  let perm = map fst $ sortMC (zip arr_layout dimAccesses)
+
+  -- Check if any of the dependencies are too complex to reason about
+  let dimAccesses' = filter (isJust . originalVar) dimAccesses
+  let deps = mapMaybe originalVar dimAccesses'
+  let counters = concatMap (map (isCounter . varType . snd) . M.toList . dependencies) dimAccesses'
+  let primExps = map (`M.lookup` primExpTable) deps
+  let inscrutable = any (uncurry isInscrutable) (zip primExps counters)
+
+  -- Check if we want to manifest this array with the permutation
+  if inscrutable || commonPermutationEliminators perm nest
+    then Nothing
+    else Just perm
+
+instance Layout MC where
+  permutationFromDimAccess = multicorePermutation
+
+instance Layout MCMem where
+  permutationFromDimAccess = multicorePermutation
+
+instance Layout GPUMem where
+  permutationFromDimAccess = error $ notImplementedYet "GPUMem"
+
+instance Layout Seq where
+  permutationFromDimAccess = error $ notImplementedYet "Seq"
+
+instance Layout SeqMem where
+  permutationFromDimAccess = error $ notImplementedYet "SeqMem"
+
+instance Layout SOACS where
+  permutationFromDimAccess = error $ notImplementedYet "SOACS"
+
+-- | Reasons common to all backends to not manifest an array.
+commonPermutationEliminators :: [Int] -> [BodyType] -> Bool
+commonPermutationEliminators perm nest = do
+  -- Don't manifest if the permutation is the permutation is invalid
+  let is_invalid_perm =
+        -- Don't manifest if the permutation is the identity permutation
+        perm `L.isPrefixOf` [0 ..]
+          -- or is not a transpose.
+          || (isNothing . isMapTranspose) perm
+          -- or is not a permutation.
+          || not (L.sort perm `L.isPrefixOf` [0 ..])
+          -- or if the last idx remains last
+          || (last perm == length perm - 1)
+
+  -- Don't manifest if the array is defined inside a segOp or loop body
+  let inside_undesired = any undesired nest
+
+  is_invalid_perm || inside_undesired
+  where
+    undesired :: BodyType -> Bool
+    undesired bodyType = case bodyType of
+      SegOpName _ -> True
+      LoopBodyName _ -> True
+      _ -> False
+
+-- | Given an ordering function for `DimAccess`, and an IndexTable, return
+-- a PermutationTable.
+-- We remove entries with no results after `permutationFromDimAccess`
+permutationTableFromIndexTable :: (Layout rep) => PrimExpTable -> IndexTable rep -> PermutationTable
```

Comment:
This is a bad name.  It's not just a mechanical translation of one data structure to another.  Is this not *the* function that decides on how to optimise the program?  It should be called something that suggests that it computes optimal permutations for the arrays, given the information in the index table.